source/common/upstream/edf_scheduler.h

```diff
@@ -0,0 +1,101 @@
+#include <cstdint>
+#include <queue>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Upstream {
+
+// It's not sufficient to use trace level logging, since it becomes far too noisy for a number of
+// tests, so we can kill trace debug here.
+#define EDF_DEBUG 0
+
+#if EDF_DEBUG
+#define EDF_TRACE(fmt...) ENVOY_LOG_MISC(trace, fmt)
+#else
+#define EDF_TRACE(fmt...)
+#endif
+
+// Earliest Deadline First (EDF) scheduler
+// (https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) used for weighted round robin.
+// Each pick from the schedule has the earliest deadline entry selected. Entries have deadlines set
+// at current time + 1 / weight, providing weighted round robin behavior with floating point
+// weights and an O(log n) pick time.
+template <class C> class EdfScheduler {
+public:
+  /**
+   * Pick queue entry with closest deadline.
+   * @return std::shared_ptr<C> to the queue entry if a valid entry exists in the queue, nullptr
+   *         otherwise. The entry is removed from the queue.
+   */
+  std::shared_ptr<C> pick() {
+    EDF_TRACE("Queue pick: queue_.size()={}, current_time_={}.", queue_.size(), current_time_);
+    while (true) {
+      if (queue_.empty()) {
+        EDF_TRACE("Queue is empty.");
+        return nullptr;
+      }
+      const EdfEntry& edf_entry = queue_.top();
+      // Entry has been removed, let's see if there's another one.
+      if (edf_entry.entry_.expired()) {
+        EDF_TRACE("Entry has expired, repick.");
+        queue_.pop();
+        continue;
+      }
+      std::shared_ptr<C> ret{edf_entry.entry_};
+      ASSERT(edf_entry.deadline_ >= current_time_);
+      current_time_ = edf_entry.deadline_;
+      order_offset_ = 0;
+      queue_.pop();
+      EDF_TRACE("Picked {}, current_time_={}.", static_cast<const void*>(ret.get()), current_time_);
+      return ret;
+    }
+  }
+
+  /**
+   * Insert entry into queue with a given weight. The deadline will be current_time_ + 1 / weight.
+   * @param weight integer weight.
+   * @param entry shared pointer to entry, only a weak reference will be retained.
+   */
+  void add(uint64_t weight, std::shared_ptr<C> entry) {
+    ASSERT(weight > 0);
+    const double deadline = current_time_ + order_offset_ + 1.0 / weight;
+    EDF_TRACE("Insertion {} in queue with deadline {} and weight {}.",
+              static_cast<const void*>(entry.get()), deadline, weight);
+    queue_.push({deadline, entry});
+    // Increment offset to ensure we get deterministic ordering. We need this bias to be small
+    // enough to avoid conflicting with other entries in the queue.
+    // TODO(htuch): this is a pretty unreliable use of floating point, probably better
```

Comment:
Should we just switch to integers given this?

Reply:
I think things get a little harder to understand using integer math. Could make the order offset an integer and store it in the EdfEntry to be used in `EdfEntry::operator<` if entries with equal deadlines are encountered.

After the initial adds during refresh, the offset never seems to be non-zero because we alternate calling pick() and add(). So this really only seems to exist for the initial ordering to be deterministic within a group of hosts with the same weight.

Reply:
Yes, this is a good idea, will do.